### DISCUSSION TITLE
Nonlinear Schrodinger equation

### DISCUSSION THREAD

**Discussion by GenB31415** [zwicker-group/py-pde#85] [Show and tell]
I used the py-pde package (author David Zwicker) to numerically investigate the dynamics of solutions (solitonic and decaying) to the nonlinear 1D Schrödinger equation (NSE). As the base schema I started from the pde_schroedinger.py that is extended by including the non-linear part, see en.wikipedia.org/wiki/Nonlinear_Schrodinger_equation. The attached files correctly describe the expected dynamic.
![DecayingState](https://user-images.githubusercontent.com/53163373/114961969-7e076f00-9e2f-11eb-88ca-d58afe34b043.png)
![DecayingState2](https://user-images.githubusercontent.com/53163373/114961972-7ea00580-9e2f-11eb-9311-7a832d15ce22.png)
![SolitonState](https://user-images.githubusercontent.com/53163373/114961976-819af600-9e2f-11eb-9414-154dc0599212.png)
![SolitonState2](https://user-images.githubusercontent.com/53163373/114961978-82cc2300-9e2f-11eb-871d-97503eec7524.png)

**Comment by GenB31415** [comment]
In this post I show the numerical solution (with py-pde) of 1D nonlinear Schrödinger equation with Double-well potential (see //en.wikipedia.org/wiki/Double-well_potential).
![Two-Well_potential1](https://user-images.githubusercontent.com/53163373/115587941-d9a08500-a293-11eb-8584-a8e554381c44.png)
![Two-Well_potential2](https://user-images.githubusercontent.com/53163373/115587944-dad1b200-a293-11eb-8a13-7f5b333d973e.png)
![Two-Well_potentialV1](https://user-images.githubusercontent.com/53163373/115587949-db6a4880-a293-11eb-941c-3eba6c874e5c.png)
![Two-Well_potentialV2](https://user-images.githubusercontent.com/53163373/115587951-dc02df00-a293-11eb-80ed-e472050432dd.png). I will show the *.py files soon.

**Reply by Elliotepsteino** [reply]
Can you please post the code used to solve the nonlinear schrödinger equation?

**Reply by GenB31415** [reply]
The required code is attached. It corresponds to the soliton dynamics for NSchEq in a double-well potential V(x).  
In line 32 parameter g = 5  can be variated: g=5>0 is a soliton state, g=-5<0 is a decaying state. Other parameters of the potential can be variated too.

Nonlinear Schrödinger's Equation, 1D case 
======================

This example implements a complex PDE using the :class:`~pde.pdes.pde.PDE`. We here
chose the `Schrödinger equation <https://en.wikipedia.org/wiki/Schrödinger_equation>`_ 
and nonlinear Schredinger equation <https://en.wikipedia.org/wiki/Nonlinear_Schr%C3%B6dinger_equation]>,
with a spatial potential V_0.V(x) in non-dimensional form:

.. math::
    i \partial_t \psi = -\nabla^2 \psi - g |\psi|^2 \psi - V_0 V(x) \psi
    
Note that the example imposes Neumann conditions at the wall, so the wave packet is
expected to reflect off the wall.
Cases 
(i) the soliton  solution(g>0, waveguide propagation), and
(ii) decaying solution (g<0) for Nonlinear Schrodinger equation.
"""
```

# slightly extended variant of linear (and potential free) case (g=0), see pde_schroedinger.py, by David Zwicker

from math import sqrt, cosh
import numpy as np
from pde import PDE, CartesianGrid, MemoryStorage, ScalarField, plot_kymograph
import matplotlib.pyplot as plt

grid = CartesianGrid([[0, 2*20]], 128, periodic=False)  # generate grid

# create a (normalized) wave packet with a certain form as an initial condition
center = grid.axes_bounds[0][1]/2
iVel=0.9                 # can be changed
g = 5                    # can be variated: g=5>0 is a soliton state, g=-5<0 is a decaying state
ampPot_V0 = 0.15          # V_0 is amplitude of potential V_0*V(x)

sInitState = "exp(I * "+str(iVel)+" * x) * exp(-(x - " + str(center) + ")**2)"
#sInitState = "exp(I * "+str(iVel)+" * x) / cosh(x - " + str(center)+ ")" # variant with sch
print("\n\nInit state: "+ sInitState)
initial_state = ScalarField.from_expression(grid, sInitState)
initial_state /= sqrt(initial_state.to_scalar("norm_squared").integral.real)

#V_x = f""+str(ampPot_V0)+"*((x-"+str(center)+")**2)" # V_0 * V(x) = V0*(x-c)^2
V_x = f""+str(ampPot_V0)+"*(-10*(x-"+str(center)+")**2 + 0.03*(x-"+str(center)+")**4)" # V_0 * V(x) = V0*(x-c)^2, OK!
#V_x = f"t*"+str(ampPot_V0)+"*(-10*(x-"+str(center)+")**2 + 0.03*(x-"+str(center)+")**4)" # V_0 * V(x) = V0*(x-c)^2
#----------plot potential----
def plot2D(*args):
    plt.grid(1)
    xAxis = np.arange(args[1],args[2],args[3])
    def xfunction(x,input):
        return eval(input)    
    plt.title(args[4])
    plt.plot(xAxis,xfunction(xAxis,args[0]))
    plt.xlabel("x"); plt.ylabel("V(x,t)")
    plt.show(block=False)

print("Plot potential: V(x)=" + V_x)  # V_0 * V(x)
plot2D(V_x, grid.axes_bounds[0][0], grid.axes_bounds[0][1], 0.1, "V(x)=" + V_x)
#--------------
sFun0 = f"I*(laplace(ψ)+("+str(g)+")*conjugate(ψ)*ψ**2)" # define the pde
sFun = f"{sFun0}+I*({V_x})*ψ"   # add potential V(x)
eq = PDE({"ψ": sFun})           # define the pde

# solve the pde and store intermediate data
storage = MemoryStorage()
eq.solve(initial_state, t_range=4*2, dt=1e-5, tracker=[storage.tracker(0.02)])

# visualize the results as a space-time plot
sFun = sFun.replace("conjugate(ψ)*ψ**2","ψ|ψ|^2") # simplify the math in title
sFun = sFun.replace("laplace","Δ")
sFun = sFun.replace("**","^")
sTitle = "dψ/dt=" + sFun + ", iV=%g" % (iVel)+ ", cnt=%g" % (center)
plot_kymograph(storage, scalar="norm_squared", title=sTitle)
print("Done.\n\n")
```

**Comment by GenB31415** [comment]
Hi Elliotepsteino
The required code is attached. It corresponds to the soliton dynamics for NSchEq in a double-well potential V(x).  
In line 32 parameter g = 5  can be variated: g=5>0 is a soliton state, g=-5<0 is a decaying state.Other parameters of the potential can be variated as well.
Regards    On Sunday, April 3, 2022, 11:18:50 PM CDT, Elliotepsteino ***@***.***> wrote:  
 
 


Can you please post the code used to solve the nonlinear schrödinger equation?

**Reply by GenB31415** [reply]
The required code is attached. It corresponds to the soliton dynamics for NSchEq in a double-well potential V(x).  
In line 32 parameter g = 5  can be variated: g=5>0 is a soliton state, g=-5<0 is a decaying state. Other parameters of the potential can be variated too.
```
Nonlinear Schrödinger's Equation, 1D case 
======================

This example implements a complex PDE using the :class:`~pde.pdes.pde.PDE`. We here
chose the `Schrödinger equation <https://en.wikipedia.org/wiki/Schrödinger_equation>`_ 
and nonlinear Schrödinger equation <https://en.wikipedia.org/wiki/Nonlinear_Schr%C3%B6dinger_equation]>,
with a spatial potential V_0.V(x) in non-dimensional form:

.. math::
    i \partial_t \psi = -\nabla^2 \psi - g |\psi|^2 \psi - V_0 V(x) \psi
    
Note that the example imposes Neumann conditions at the wall, so the wave packet is
expected to reflect off the wall.
Cases 
(i) the soliton  solution(g>0, waveguide propagation), and
(ii) decaying solution (g<0) for Nonlinear Schrodinger equation.
"""

# slightly extended variant of linear (and potential free) case (g=0), see pde_schroedinger.py, by David Zwicker

from math import sqrt, cosh
import numpy as np
from pde import PDE, CartesianGrid, MemoryStorage, ScalarField, plot_kymograph
import matplotlib.pyplot as plt

grid = CartesianGrid([[0, 2*20]], 128, periodic=False)  # generate grid

# create a (normalized) wave packet with a certain form as an initial condition
center = grid.axes_bounds[0][1]/2
iVel=0.9                 # can be changed
g = 5                    # can be variated: g=5>0 is a soliton state, g=-5<0 is a decaying state
ampPot_V0 = 0.15          # V_0 is amplitude of potential V_0*V(x)

sInitState = "exp(I * "+str(iVel)+" * x) * exp(-(x - " + str(center) + ")**2)"
#sInitState = "exp(I * "+str(iVel)+" * x) / cosh(x - " + str(center)+ ")" # variant with sch
print("\n\nInit state: "+ sInitState)
initial_state = ScalarField.from_expression(grid, sInitState)
initial_state /= sqrt(initial_state.to_scalar("norm_squared").integral.real)

#V_x = f""+str(ampPot_V0)+"*((x-"+str(center)+")**2)" # V_0 * V(x) = V0*(x-c)^2
V_x = f""+str(ampPot_V0)+"*(-10*(x-"+str(center)+")**2 + 0.03*(x-"+str(center)+")**4)" # V_0 * V(x) = V0*(x-c)^2, OK!
#V_x = f"t*"+str(ampPot_V0)+"*(-10*(x-"+str(center)+")**2 + 0.03*(x-"+str(center)+")**4)" # V_0 * V(x) = V0*(x-c)^2
#----------plot potential----
def plot2D(*args):
    plt.grid(1)
    xAxis = np.arange(args[1],args[2],args[3])
    def xfunction(x,input):
        return eval(input)    
    plt.title(args[4])
    plt.plot(xAxis,xfunction(xAxis,args[0]))
    plt.xlabel("x"); plt.ylabel("V(x,t)")
    plt.show(block=False)

print("Plot potential: V(x)=" + V_x)  # V_0 * V(x)
plot2D(V_x, grid.axes_bounds[0][0], grid.axes_bounds[0][1], 0.1, "V(x)=" + V_x)
#--------------
sFun0 = f"I*(laplace(ψ)+("+str(g)+")*conjugate(ψ)*ψ**2)" # define the pde
sFun = f"{sFun0}+I*({V_x})*ψ"   # add potential V(x)
eq = PDE({"ψ": sFun})           # define the pde

# solve the pde and store intermediate data
storage = MemoryStorage()
eq.solve(initial_state, t_range=4*2, dt=1e-5, tracker=[storage.tracker(0.02)])

# visualize the results as a space-time plot
sFun = sFun.replace("conjugate(ψ)*ψ**2","ψ|ψ|^2") # simplify the math in title
sFun = sFun.replace("laplace","Δ")
sFun = sFun.replace("**","^")
sTitle = "dψ/dt=" + sFun + ", iV=%g" % (iVel)+ ", cnt=%g" % (center)
plot_kymograph(storage, scalar="norm_squared", title=sTitle)
print("Done.\n\n")
```
